error recovery takes long time

### Context
System version: 5.2.1
libfranka version: 0.10.0
franka_ros version: 0.10.2 (unreleased), on commit 2d458abad7390bb73771c91787c361c2b5cfa6ad.
controller: position_joint_trajectory_controller

### Problem
When an error happens (in particular cartesian reflexes as this is what we tested), the system takes a long and varying time to confirm it. This comes in two parts:

1.  There is a significant time difference between the manipulator stopping during a trajectory (we carefully blocked the movement of the robot with our hands) and the cartesian reflex being reported by libfranka. Experimentally we found this can be up to 1 or 2 seconds.
2. Once the error is reported, we identify it through the robot mode message attribute published in /franka_state_controller/franka_states. When the robot mode is not "normal" (IDLE or MOVE), we check it and confirm it. A simplified version of the script to generate the ROS node we use for this can be find below.

The time difference between the error recovery request to be sent and the error recovery varies a lot.  In an experiment where we confirmed 11 errors, we found the following time differences in seconds:

1.0152
0.9475
15.9851
2.5994
0.3952
0.8022
1.2532
0.5220
2.1730
0.6632
3.0928

It can be seen that the confirmation is sometimes very quick, other times a bit longer and rarely it takes a very long time.

Part 2 of the problem is more critical for us. 

### Steps to reproduce part 2

1. Start all necessary nodes to run the robot including the franka_control node
2. Start the manipulator_error_manager node with the script below
3. Move the robot and produce a cartesian reflex by carefully blocking the robot's movement
4. Compare the timestamps of the logs printed by libfranka libfranka: Move command aborted: motion aborted by reflex! ["cartesian_reflex"] and `Recovered from error`
5. Repeat steps 3 and 4 multiple times

### Steps to reproduce part 1
1. Start all necessary nodes to run the robot including the franka_control node
2. Start the manipulator_error_manager node with the script below
3. Move the robot and produce a cartesian reflex by carefully blocking the robot's movement
4. Observe how much time it takes for the robot light to turn red
5. Repeat steps 3 and 4 multiple times

### Script of the node to confirm errors
```
#!/usr/bin/python3

import rospy
import traceback
import actionlib
import franka_msgs.msg as franka_msg

class ManipulatorErrorManager:
    '''
    class containing code the get the status of the panda manipulator and chnage this state (mainly confirming errors)
    '''
    def __init__(self):
        # Robot modes and their meaning:
        self.robot_modes_meaning = {'0': 'OTHER',
                                    '1': 'IDLE',
                                    '2': 'MOVE',
                                    '3': 'GUIDING',
                                    '4': 'REFLEX',
                                    '5': 'USER_STOPPED',
                                    '6': 'AUTOMATIC_ERROR_RECOVERY'}
        
        self.confirmation_cooldown_time = 0.2 # [s] cooldown time after an error confirmation attempt

        # Init attributes        
        self.confirmation_time_stamp = rospy.Time.now() # time stamp the of the last confirmation attempt

        # Init subscriber to franka state 
        rospy.Subscriber('/franka_state_controller/franka_states', franka_msg.FrankaState, self.__callback_status_information, queue_size=1)

        # Initialize action client for error recovery
        self.error_recovery_client = actionlib.SimpleActionClient('/franka_control/error_recovery', franka_msg.ErrorRecoveryAction)
        self.error_recovery_client.wait_for_server()
    
    def __callback_status_information(self, msg: franka_msg.FrankaState) -> None:
        ''' Callback method of the subscriber to the franka status topic
        - @param msg: the message that was send by the franka state interface 
        Documentation about the status message: https://frankaemika.github.io/libfranka/structfranka_1_1RobotState.html
        '''
        duration_since_last_confirmation = (rospy.Time.now() - self.confirmation_time_stamp).to_sec()

        # If robot has an error and if not in cooldown time, confirm the error
        if not msg.robot_mode in [1, 2, 5] and duration_since_last_confirmation > self.confirmation_cooldown_time:
            robot_mode_description = self.robot_modes_meaning[str(msg.robot_mode)]
            rospy.logwarn(f"Manipulator error detected. Robot mode: {robot_mode_description}")

            # Define action message
            goal_msg = franka_msg.ErrorRecoveryActionGoal()

            # Trigger action server -> confirm error
            self.error_recovery_client.send_goal_and_wait(goal_msg)

            # Reset confirmation time stamp
            self.confirmation_time_stamp = rospy.Time.now()
            
            rospy.loginfo("Manipulator error confirmed")

def main():
    try:
        rospy.loginfo('Starting Manipulator Error Manager Node')

        rospy.init_node('manipulator_error_manager')

        # Create error manager object
        error_manager = ManipulatorErrorManager()

        rospy.spin()

    except (rospy.ROSInterruptException, KeyboardInterrupt):
        return

    except:
        rospy.logfatal('uncaught error in manipulator_error_manager.py')
        rospy.logfatal(traceback.format_exc())
        return

if __name__ == '__main__':
    main()
```

I did the experiment again with the [PR](https://github.com/frankaemika/franka_ros/pull/279) of #232 and unfortunately still got similar durations as the ones reported previously. Here there is a table with the new durations.

Timestamp of 1 [s] | Timestamp of 2 [s] | Timestamp of 3 [s] | Duration from 1 to 2 [s] | % | Duration from 2 to 3 [s] | % |  Duration from 1 to 3 [s]
-- | -- | -- | -- | -- | -- | -- | --
1677673485.6275 | 1677673485.62889 | 1677673492.96849 | 0.0014 | 0.02% | 7.3396 | 99.98% | 7.3410
1677673816.95238 | 1677673816.95447 | 1677673819.68224 | 0.0021 | 0.08% | 2.7278 | 99.92% | 2.7299
1677674038.47682 | 1677674038.47934 | 1677674039.17731 | 0.0025 | 0.36% | 0.6980 | 99.64% | 0.7005
1677674147.85044 | 1677674147.85261 | 1677674151.59693 | 0.0022 | 0.06% | 3.7443 | 99.94% | 3.7465
1677674269.48645 | 1677674269.48814 | 1677674272.10449 | 0.0017 | 0.06% | 2.6164 | 99.94% | 2.6180
1677674397.38669 | 1677674397.38861 | 1677674399.88418 | 0.0019 | 0.08% | 2.4956 | 99.92% | 2.4975
1677674508.8368 | 1677674508.83859 | 1677674509.58911 | 0.0018 | 0.24% | 0.7505 | 99.76% | 0.7523
1677674643.74671 | 1677674643.74849 | 1677674644.79248 | 0.0018 | 0.17% | 1.0440 | 99.83% | 1.0458
1677674731.5977 | 1677674731.60386 | 1677674732.48572 | 0.0062 | 0.69% | 0.8819 | 99.31% | 0.8880
1677674823.51982 | 1677674823.52105 | 1677674823.68413 | 0.0012 | 0.75% | 0.1631 | 99.25% | 0.1643
average |   |   | 0.0023 | 0.26% | 2.2850 | 99.74% | 2.2874

In case it is useful, I noticed that the robot first stops, the LEDs stay in green for some time, then they change to red and quickly to green again.

This were the logs for the first attempt of the experiment just as an example of how the logs looked. I took the durations from these timestamps.
```
[ERROR] [1677673485.627496941]: libfranka: Move command aborted: motion aborted by reflex! ["joint_reflex", "cartesian_reflex"]
control_command_success_rate: 1
[ INFO] [1677673485.627570161]: franka_control: main loop
[WARN] [1677673485.628891]: Manipulator error detected. Robot mode: REFLEX
[ INFO] [1677673492.968491856]: Recovered from error
[ INFO] [1677673492.968849281]: franka_control: controller activated
[INFO] [1677673492.971180]: Manipulator error confirmed
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Requires `libfranka` >= 0.8.0
   * `franka_gazebo`: Offer both `panda.launch` and `fr3.launch`.
   * `franka_gazebo`: Fix velocity control by adding the missing effort.
   * `franka_control`: Clear the error flag if the robot is in `kIdle` mode, i.e. ready to move.
+  * `franka_control`: Fix unexpected long delay when calling error recovery ([#317](https://github.com/frankaemika/franka_ros/issues/317))
   * Fix a possible compilation error by sorting include directories by topological order ([#319](https://github.com/frankaemika/franka_ros/issues/319)).
 
 ## 0.10.1 - 2022-09-15

--- a/franka_control/src/franka_control_node.cpp
+++ b/franka_control/src/franka_control_node.cpp
@@ -130,6 +130,7 @@ int main(int argc, char** argv) {
           }
         } catch (const std::logic_error& e) {
         }
+        std::this_thread::yield();
       } else {
         std::this_thread::sleep_for(1ms);
       }

